r2dbc-core/src/main/java/com/google/cloud/sql/core/GcpConnectionFactoryProvider.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2020 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.sql.core;
+
+import static io.r2dbc.spi.ConnectionFactoryOptions.Builder;
+import static io.r2dbc.spi.ConnectionFactoryOptions.DRIVER;
+import static io.r2dbc.spi.ConnectionFactoryOptions.HOST;
+import static io.r2dbc.spi.ConnectionFactoryOptions.PROTOCOL;
+
+import io.netty.handler.ssl.SslContextBuilder;
+import io.r2dbc.spi.ConnectionFactory;
+import io.r2dbc.spi.ConnectionFactoryOptions;
+import io.r2dbc.spi.ConnectionFactoryProvider;
+import io.r2dbc.spi.Option;
+import java.util.Properties;
+import java.util.function.Function;
+
+/** {@link ConnectionFactoryProvider} for proxied access to GCP Postgres and MySQL instances. */
+public abstract class GcpConnectionFactoryProvider implements ConnectionFactoryProvider {
+
+  abstract ConnectionFactory tcpConnectonFactory(
```

Comment:
It would be great if we could get some brief comments on these interfaces - they make sense now, but I'm worried in a month or two I may forget their importance.

Reply:
done